state/migrations/20231108122539_clear_stuck_invites_test.go

```diff
@@ -0,0 +1,212 @@
+package migrations
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/matrix-org/sliding-sync/sqlutil"
+	"github.com/matrix-org/sliding-sync/state"
+	"github.com/matrix-org/sliding-sync/sync2"
+)
+
+func TestClearStuckInvites(t *testing.T) {
+	db, close := connectToDB(t)
+	defer close()
+	roomsTable := state.NewRoomsTable(db)
+	inviteTable := state.NewInvitesTable(db)
+	unreadTable := state.NewUnreadTable(db)
+	deviceTable := sync2.NewDevicesTable(db)
+	tokensTable := sync2.NewTokensTable(db, "secret")
+
+	zero := 0
+	device1 := "TEST_1"
+	device2 := "TEST_2"
+	roomA := "!TestClearStuckInvites_a:localhost"
+	roomB := "!TestClearStuckInvites_b:localhost"
+	roomC := "!TestClearStuckInvites_c:localhost"
+	roomD := "!TestClearStuckInvites_d:localhost"
+	roomE := "!TestClearStuckInvites_e:localhost"
+	roomF := "!TestClearStuckInvites_f:localhost"
+	roomG := "!TestClearStuckInvites_g:localhost"
+	alice := "@TestClearStuckInvites_alice:localhost"
+	bob := "@TestClearStuckInvites_bob:localhost"
+	charlie := "@TestClearStuckInvites_charlie:localhost"
+	doris := "@TestClearStuckInvites_doris:localhost"
+	users := []string{
+		alice, bob, charlie, doris,
+	}
+
+	// Test cases:
+	// Room | In Invite Table? | In Unread Table? | In Room Table? | Comment
+	//   A           Y                 Y                Y             OK, Genuine invite, proxy in room
+	//   B           Y                 Y                N             BAD, Stuck invite, proxy never saw join
+	//   C           Y                 N                Y             OK, Genuine invite, proxy in room, no unread counts (unusual but valid)
+	//   D           Y                 N                N             OK, Genuine invite, proxy not in room
+	//   E           N                 Y                Y             OK, Genuine joined room
+	//   F           N                 Y                N             BAD, Stuck joined room, proxy never saw join
+	//   G           N                 N                Y             OK, Genuine joined room, no unread counts (unusual but valid)
+	//   -           N                 N                N             Impossible, room id isn't in any table!
```

Comment:
There's an assumption here that we never delete from the rooms table. Is that true? (I think it should be, but let's check)

Reply:
Yes, with the exception of a migration which did delete "bad" rooms.